### ISSUE TITLE
Numerous "could not listen..." errors after updating Android 11 device to Sept. 5 security patch

### ISSUE THREAD

**Issue by grnassar** [Genymobile/scrcpy#2624]
- [x] I have read the [FAQ](https://github.com/Genymobile/scrcpy/blob/master/FAQ.md).
 - [x] I have searched in existing [issues](https://github.com/Genymobile/scrcpy/issues).

**Environment**
 - OS: [Windows 10 21H1]
 - scrcpy version: [1.18]
 - installation method: [Windows binary, 64-bit]
 - device model: Pixel 3a
 - Android version: [11]

**Describe the bug**

Device (Android 11) was working fine with scrcpy prior to 9/5 security update.  After security update, app fails to run. adb devices seems to show the device fine (only one connected), adb shell gets a device shell fine.  Console output and adb logcat (flushed right before trying scrcpy) given below.

console output (some repetitive lines skipped):
```
D:\Programs\scrcpy-win64\scrcpy-server: 1 file pushed, 0 skipped. 30.3 MB/s (37330 bytes in 0.001s)
bind: No error
WARN: Could not listen on port 27183, retrying on 27184
bind: No error
WARN: Could not listen on port 27184, retrying on 27185
bind: No error
...
WARN: Could not listen on port 27198, retrying on 27199
bind: No error
ERROR: Could not listen on any port in range 27183:27199
WARN: 'adb reverse' failed, fallback to 'adb forward'
adb.exe: error: cannot bind listener: cannot bind to 127.0.0.1:27183: An attempt was made to access a socket in a way forbidden by its access permissions. (10013)
ERROR: "adb forward" returned with value 1
WARN: Could not forward port 27183, retrying on 27184
adb.exe: error: cannot bind listener: cannot bind to 127.0.0.1:27184: An attempt was made to access a socket in a way forbidden by its access permissions. (10013)
ERROR: "adb forward" returned with value 1
WARN: Could not forward port 27184, retrying on 27185
...
WARN: Could not forward port 27198, retrying on 27199
adb.exe: error: cannot bind listener: cannot bind to 127.0.0.1:27199: An attempt was made to access a socket in a way forbidden by its access permissions. (10013)
ERROR: "adb forward" returned with value 1
ERROR: Could not forward any port in range 27183:27199
```

logcat (don't think it's relevant, not sure):
```
--------- beginning of system
09-09 07:48:40.741  1403  1788 I chatty  : uid=1000(system) ConnectivitySer expire 18 lines
--------- beginning of main
09-09 07:48:42.694  2789  2813 I chatty  : uid=10140(com.google.android.apps.scone) StateService expire 2 lines
09-09 07:48:42.698  2789  2813 D DeviceStateHelper: Audo mode: 0
09-09 07:48:43.774  1403  1788 D ConnectivityService: NetReassign [no changes]
09-09 07:48:43.987  7231  8862 D [TORNET]: [TORNET][NOTE ] Sending Heartbeat to Server
09-09 07:48:44.188  7231  8862 D [TORNET]: [TORNET][DEBUG] Received HeartBeat from Server
09-09 07:48:44.962   905   905 I android.hardware.health@2.0-service.bonito: capacity: 3606000
09-09 07:48:44.981  1878  2993 D PowerUI : can't show warning due to - plugged: true status unknown: false
09-09 07:48:46.805  1403  1788 D ConnectivityService: NetReassign [no changes]
09-09 07:48:47.724  2789  2813 D DeviceStateHelper: Audo mode: 0
```

**Comment by rom1v** [comment]
See #2468.

**Comment by grnassar** [comment]
I saw #2468 when I searched.  I have no pending Windows updates, and have tried restarting both the host and target device multiple times to no avail.  Also, the user reported that it happened when switching from Android 10 to 11, which I did not experience (it continued working fine for me then).  It was only after the most recent security update that scrcpy broke for me.

**Comment by rom1v** [comment]
Try without scrcpy at all:
```bash
adb reverse localabstract:test tcp:27183
```

**Comment by grnassar** [comment]
> 
> 
> Try without scrcpy at all:
> 
> ```shell
> adb reverse localabstract:test tcp:27183
> ```

What is the expected result of this command?
I don't get any console output when I run it; it just returns me to the command prompt.

**Comment by grnassar** [comment]
> 
> 
> Try without scrcpy at all:
> 
> ```shell
> adb reverse localabstract:test tcp:27183
> ```

What is the expected result of this command?
I don't get any console output when I run it; it just returns me to the command prompt.

**Comment by rom1v** [comment]
If there is no error message and the redirection appears in `adb reverse --list`, then it worked.

Remove the redirection (`adb reverse --remove-all`), and retry scrcpy.

> It was only after the most recent security update that scrcpy broke for me.

The problem is on the host (Windows):

> cannot bind to 127.0.0.1:27183: An attempt was made to access a socket in a way forbidden by its access permissions. (10013)

**Comment by grnassar** [comment]
Sa

> 
> 
> If there is no error message and the redirection appears in `adb reverse --list`, then it worked.
It shows up in adb reverse --list, yes.

> Remove the redirection (`adb reverse --remove-all`), and retry scrcpy.

Still fails, same output.
> > It was only after the most recent security update that scrcpy broke for me.

> The problem is on the host (Windows):
> 
> > cannot bind to 127.0.0.1:27183: An attempt was made to access a socket in a way forbidden by its access permissions. (10013)

Please note that the quoted error is *not* the initial problem.
That quoted error occurs *only after* the following:
```
ERROR: Could not listen on any port in range 27183:27199
WARN: 'adb reverse' failed, fallback to 'adb forward'
```

The adb reverse connect method for scrcpy worked fine prior to this security update.  I never needed to use the adb forward fallback to begin with.

**Comment by rom1v** [comment]
> Please note that the quoted error is not the initial problem.

OK:

> WARN: Could not listen on port 27183, retrying on 27184

The error is still on the host (it could not listen on 27183, and all other ports).

**Comment by grnassar** [comment]
As you saw, adb reverse works fine in isolation.

Posting full console log of previous requested commands for confirmation.

```
PS D:\Programs\scrcpy-win64> ./adb reverse --list

PS D:\Programs\scrcpy-win64> ./adb reverse localabstract:test tcp:27183
PS D:\Programs\scrcpy-win64> ./adb reverse --list
UsbFfs localabstract:test tcp:27183

PS D:\Programs\scrcpy-win64> ./adb reverse --remove-all
PS D:\Programs\scrcpy-win64> ./adb reverse --list

PS D:\Programs\scrcpy-win64> ./scrcpy
INFO: scrcpy 1.18 <https://github.com/Genymobile/scrcpy>
D:\Programs\scrcpy-win64\scrcpy-server: 1 file pushed, 0 skipped. 43.2 MB/s (37330 bytes in 0.001s)
bind: No error
WARN: Could not listen on port 27183, retrying on 27184
bind: No error
WARN: Could not listen on port 27184, retrying on 27185
...
bind: No error
ERROR: Could not listen on any port in range 27183:27199
WARN: 'adb reverse' failed, fallback to 'adb forward'
adb.exe: error: cannot bind listener: cannot bind to 127.0.0.1:27183: An attempt was made to access a socket in a way forbidden by its access permissions. (10013)
ERROR: "adb forward" returned with value 1
WARN: Could not forward port 27183, retrying on 27184
adb.exe: error: cannot bind listener: cannot bind to 127.0.0.1:27184: An attempt was made to access a socket in a way forbidden by its access permissions. (10013)
ERROR: "adb forward" returned with value 1
WARN: Could not forward port 27184, retrying on 27185
adb.exe: error: cannot bind listener: cannot bind to 127.0.0.1:27185: An attempt was made to access a socket in a way forbidden by its access permissions. (10013)
ERROR: "adb forward" returned with value 1
...
WARN: Could not forward port 27198, retrying on 27199
adb.exe: error: cannot bind listener: cannot bind to 127.0.0.1:27199: An attempt was made to access a socket in a way forbidden by its access permissions. (10013)
ERROR: "adb forward" returned with value 1
ERROR: Could not forward any port in range 27183:27199
```

**Comment by rom1v** [comment]
OK, so it seems your system lets `adb` bind sockets, but does not let `scrcpy` do so:

https://github.com/Genymobile/scrcpy/blob/c96f5c70e9c6fe82da059c7d24e77d9730aa7f86/app/src/util/net.c#L63-L64

Maybe some firewall?

**Comment by grnassar** [comment]
Default Windows firewall has been in use the whole time, including prior to the security update when it worked fine.  No third-party software that would add a firewall or manage Windows' default firewall, neither before nor added after the security update.  Disabling Windows firewall seems to have no effect.

**Comment by grnassar** [comment]
Also seems interesting that perror() is returning "No error".

**Comment by rom1v** [comment]
I don't know why your system refuses to `bind()` from scrcpy.

> Also seems interesting that perror() is returning "No error".

Yes, it always does this on Windows apparently, no idea why.

**Comment by grnassar** [comment]
> EDIT: because Windows uses a different function to get the error: `WSAGetLastError()` https://stackoverflow.com/a/15797381/1987178

Looks like ifdef'ing the error function to give a usable error under Windows may be useful, since as you posted earlier this problem of perror() not returning anything usable has come up before.

**Comment by grnassar** [comment]
> 
> 

> 
> > Also seems interesting that perror() is returning "No error".
> 
> Yes, it always does this on Windows apparently, no idea why.
> 
> EDIT: because Windows uses a different function to get the error: `WSAGetLastError()` https://stackoverflow.com/a/15797381/1987178

**Comment by rom1v** [comment]
> Looks like ifdef'ing the error function (or however you prefer to differentiate target platforms) to give a usable error under Windows may be useful, since as you posted earlier this problem of perror() not returning anything usable has come up before.

Yes.

I suggest a `sc_perror()` which calls `perror()` on all systems except Windows, and Windows-specific stuff on Windows.

Would you like to work on it and submit a PR?

**Comment by grnassar** [comment]
Sorry, not exactly sure how I accidentally closed that earlier.

I'm a bit swamped with work, but given time if nobody else gets to it for a while I could give it a shot.

**Comment by rom1v** [comment]
OK, so I implemented it: 8fb6f09fffb5b0d3ffde798a1e99ec50f54be4d1

Could you try before I merge into `dev`, please?

Here is a binary (to replace in your v1.18 release):
[`scrcpy.exe`](https://tmp.rom1v.com/scrcpy/2624/1/scrcpy.exe) _sha256:bd0aed821d0ddffeef8e114f63671da26ec14ec4c51eefd6d306f2bfed748c54_

**Comment by grnassar** [comment]
I'm glad to hear you say that :) I went ahead and tried quickly setting up the build environment, but ran into a few problems doing so.  (Probably my fault; it's been a while since I've used Debian, plus I'm trying it under WSL2 which probably brings its own complications, no matter how compatible they advertise it as...)

I'll give this a shot and let you know how it looks here within the hour.

**Comment by grnassar** [comment]
Looks like the change worked.  Perhaps one minor cosmetic thing needs fixing, only to make it consistent with the error reporting from the unix-style sockets.  See below:

```
INFO: scrcpy 1.18 <https://github.com/Genymobile/scrcpy>
D:\Programs\scrcpy-win64\scrcpy-server: 1 file pushed, 0 skipped. 3.5 MB/s (37330 bytes in 0.010s)
bind: [10013] An attempt was made to access a socket in a way forbidden by its access permissions.

WARN: Could not listen on port 27183, retrying on 27184
bind: [10013] An attempt was made to access a socket in a way forbidden by its access permissions.

WARN: Could not listen on port 27184, retrying on 27185
...
bind: [10013] An attempt was made to access a socket in a way forbidden by its access permissions.

WARN: Could not listen on port 27198, retrying on 27199
bind: [10013] An attempt was made to access a socket in a way forbidden by its access permissions.

ERROR: Could not listen on any port in range 27183:27199
WARN: 'adb reverse' failed, fallback to 'adb forward'
adb.exe: error: cannot bind listener: cannot bind to 127.0.0.1:27183: An attempt was made to access a socket in a way forbidden by its access permissions. (10013)
ERROR: "adb forward" returned with value 1
WARN: Could not forward port 27183, retrying on 27184
adb.exe: error: cannot bind listener: cannot bind to 127.0.0.1:27184: An attempt was made to access a socket in a way forbidden by its access permissions. (10013)
ERROR: "adb forward" returned with value 1
WARN: Could not forward port 27184, retrying on 27185
...
adb.exe: error: cannot bind listener: cannot bind to 127.0.0.1:27199: An attempt was made to access a socket in a way forbidden by its access permissions. (10013)
ERROR: "adb forward" returned with value 1
ERROR: Could not forward any port in range 27183:27199
```

As you can see, the new error message adds an extra carriage return at the end, which probably shouldn't be there for consistency's sake.  Like I said -- very minor issue, and a very useful change!

Of course, troubleshooting the problem I'm having remains a mystery. :)  I'd done some looking while I was waiting for some Debian packages to install, and the last reason listed here caught my eye: https://www.easytechjunkie.com/what-are-the-common-causes-of-socket-error-10013.htm Seems like a bit of a longshot, but I've already tried most suggestions, so why not. They mention a driver being bound to multiple devices could cause it; I'm wondering if post-security update the device identified itself as a "new device" and there are some phantom copies of this device in Device Manager's "hidden devices."  That, and another page suggested manually resetting Winsock and the IP stack before rebooting.  I'm going to try both of those; I'll let you know if I have success with either.  Hopefully we get this resolved too!  Thank you.

**Comment by rom1v** [comment]
> As you can see, the new error message adds an extra carriage return at the end

Good catch :+1:

The String returned by `FormatMessage()` already contains a newline character: https://devblogs.microsoft.com/oldnewthing/20191025-00/?p=103025 :roll_eyes: 

Fixed: 1d1c9f36f43314523187b8e1fb868d03718dcc79

**Comment by grnassar** [comment]
Didn't get a chance to test the second idea, because the first one (the one from the link) worked.  After disconnecting the device, I went to Device Manager, selected the "show hidden devices" option in the View menu, found the device's old entry in the Portable Devices category and uninstalled it.  (Had I found multiples, I would've uninstalled them all, but I only found one.)  Rebooted, tried again and scrcpy worked just fine.  Hopefully that helps the next person that runs into the problem!
